src/common/ceph_json.h

```diff
@@ -467,4 +467,130 @@ void encode_json_map(const char *name, const char *index_name, const char *value
   encode_json_map<K, V>(name, index_name, NULL, value_name, NULL, NULL, m, f);
 }
 
+struct JSONFormattable {
+  enum Type {
+    FMT_NONE,
+    FMT_STRING,
+    FMT_ARRAY,
+    FMT_OBJ,
+  } type{FMT_NONE};
+
+  string str;
+  vector<JSONFormattable> arr;
+  map<std::string, JSONFormattable> obj;
+  void decode_json(JSONObj *jo) {
+    if (jo->is_array()) {
+      type = JSONFormattable::FMT_ARRAY;
+      decode_json_obj(arr, jo);
+    } else if (jo->is_object()) {
+      type = JSONFormattable::FMT_OBJ;
+      auto iter = jo->find_first();
+      for (;!iter.end(); ++iter) {
+        JSONObj *field = *iter;
+        decode_json_obj(obj[field->get_name()], field);
+      }
+    } else {
+      type = JSONFormattable::FMT_STRING;
+      decode_json_obj(str, jo);
+    }
+  }
+
+  void encode(bufferlist& bl) const {
+    ENCODE_START(1, 1, bl);
+    encode((uint8_t)type, bl);
+    encode(str, bl);
+    encode(arr, bl);
+    encode(obj, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::iterator& bl) {
+    DECODE_START(1, bl);
+    uint8_t t;
+    decode(t, bl);
+    type = (Type)t;
+    decode(str, bl);
+    decode(arr, bl);
+    decode(obj, bl);
+    DECODE_FINISH(bl);
+  }
+
+  const std::string& val() const {
+    return str;
+  }
+
+  int val_int() const;
+  bool val_bool() const;
+
+  const vector<JSONFormattable>& array() const {
+    return arr;
+  }
+
+  const JSONFormattable& operator[](const std::string& name) const;
+  const JSONFormattable& operator[](size_t index) const;
+
+  JSONFormattable& operator[](const std::string& name);
+  JSONFormattable& operator[](size_t index);
+
+  operator std::string() const {
+    return str;
+  }
+
+  explicit operator int() const {
+    return val_int();
+  }
+
+  explicit operator bool() const {
+    return val_bool();
+  }
+
+  template<class T>
+  T operator[](const std::string& name) const {
+    return this->operator[](name)(T());
+  }
+
+  template<class T>
+  T operator[](const std::string& name) {
+    return this->operator[](name)(T());
+  }
+
+  string operator ()(const char *def_val) const {
+    return def(string(def_val));
+  }
```

Comment:
i don't think this overload is needed - `operator()("foo")` will already call the `const string&` overload because `const char*` implicitly converts to `string`